File: 36_Flip_Flop/ruby/flipflop.rb
#A class representing the internal state of a single game of flip flop
# state represents the list of X's (A in the original code)
# guesses represents the number of guesses the user has made (C in the original code)
# seed represents the random seed for an instance of the game (Q in the original code)
Game = Struct.new(:state, :guesses, :seed) do

  #The original BASIC program used 1 indexed arrays while Ruby has 0-indexed arrays.
  #We can't use 0 indexed arrays for the flip functions or we'll get divide by zero errors.
  #These convenience functions allow us to modify and access internal game state in a 1-indexed fashion
  def flip_letter(letter_number)
    index = letter_number -1
    if self.state[index] == 'X'
      self.state[index] = 'O'
    else
      self.state[index] = 'X'
    end
  end

  def letter_at(letter_number)
    self.state[letter_number - 1]
  end
end

def print_welcome
  puts 'FLIPFLOP'.center(72)
  puts 'CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY'.center(72)
  puts <<~EOS

    THE OBJECT OF THIS PUZZLE IS TO CHANGE THIS:

    X X X X X X X X X X

    TO THIS:

    O O O O O O O O O O

    BY TYPING THE NUMBER CORRESPONDING TO THE POSITION OF THE
    LETTER ON SOME NUMBERS, ONE POSITION WILL CHANGE, ON
    OTHERS, TWO WILL CHANGE.  TO RESET LINE TO ALL X'S, TYPE 0
    (ZERO) AND TO START OVER IN THE MIDDLE OF A GAME, TYPE
    11 (ELEVEN).
  EOS
end

def print_starting_message
  puts <<~EOS

  HERE IS THE STARTING LINE OF X'S.

  1 2 3 4 5 6 7 8 9 10
  X X X X X X X X X X
  
  EOS
end

#Create a new game with [X,X,X,X,X,X,X,X,X,X] as the state
#0 as the number of guesses and a random seed between 0 and 1
def generate_new_game
  Game.new(Array.new(10, 'X'), 0, rand())
end

#Given a game, an index, and a shuffle function, flip one or more letters
def shuffle_board(game, index, shuffle_function)
  n = method(shuffle_function).call(game, index)

  if game.letter_at(n) == "O"
    game.flip_letter(n)
    if index == n
      n = shuffle_board(game, index, shuffle_function)
    end
  else
    game.flip_letter(n)
  end
  return n
end

#Shuffle logic copied from original BASIC code
def shuffle_function1(game, index) 
  r = Math.tan(game.seed + index / game.seed - index) - Math.sin(game.seed / index) + 336 * Math.sin(8 * index)
  n = r - r.floor
  (10 * n).floor
end

def shuffle_function2(game, index)
  r = 0.592 * (1/ Math.tan(game.seed / index + game.seed)) / Math.sin(index * 2 + game.seed) - Math.cos(index)
  n = r - r.floor
  (10 * n)
end

def play_game
  print_starting_message
  game = generate_new_game
  working_index = nil

  loop do
    puts "INPUT THE NUMBER"
    input = gets.chomp.downcase

    #See if the user input a valid integer, fail otherwise
    if numeric_input = Integer(input, exception: false)

      #If 11 is entered, we're done with this version of the game
      if numeric_input == 11
        return :restart
      end

      if numeric_input > 11
        puts 'ILLEGAL ENTRY--TRY AGAIN.'
        next #illegal entries don't count towards your guesses
      end

      if working_index == numeric_input
        game.flip_letter(numeric_input)
        working_index = shuffle_board(game, numeric_input, :shuffle_function2)
      #If 0 is entered, we want to reset the state, but not the random seed or number of guesses and keep playing
      elsif numeric_input == 0
        game.state = Array.new(10, 'X')
      elsif game.letter_at(numeric_input) == "O"
        game.flip_letter(numeric_input)
        if numeric_input == working_index
          working_index = shuffle_board(game, numeric_input, :shuffle_function1)
        end
      else
        game.flip_letter(numeric_input)
        working_index = shuffle_board(game, numeric_input, :shuffle_function1)
      end
    else
      puts 'ILLEGAL ENTRY--TRY AGAIN.'
      next #illegal entries don't count towards your guesses
    end

    game.guesses += 1
    puts '1 2 3 4 5 6 7 8 9 10'
    puts game.state.join(' ')
    
    if game.state.all? { |x| x == 'O' }
      if game.guesses > 12 
        puts "TRY HARDER NEXT TIME. IT TOOK YOU #{game.guesses} GUESSES."
      else
        puts "VERY GOOD.  YOU GUESSED IT IN ONLY #{game.guesses} GUESSES."
      end
      #game is complete
      return
    end
  end
end



#Execution starts
print_welcome
loop do
  result = play_game
  if result == :restart
    next
  end

  puts 'DO YOU WANT TO TRY ANOTHER PUZZLE'
  if gets.chomp.downcase[0] == 'n'
    break
  end
end

File: 36_Flip_Flop/perl/flipflop.pl
#!/usr/bin/perl

# Flip Flop program in Perl
# Translated by Kevin Brannen (kbrannen)

use strict;
use warnings;
use Math::Trig;

print "\n";
print " " x 32, "FLIPFLOP";
print " " x 15, "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n\n\n";
# *** CREATED BY MICHAEL CASS

print "THE OBJECT OF THIS PUZZLE IS TO CHANGE THIS:\n\n";
print "X X X X X X X X X X\n\n";
print "TO THIS:\n\n";
print "O O O O O O O O O O\n\n";
print "BY TYPING THE NUMBER CORRESPONDING TO THE POSITION OF THE\n";
print "LETTER ON SOME NUMBERS, ONE POSITION WILL CHANGE, ON\n";
print "OTHERS, TWO WILL CHANGE.  TO RESET LINE TO ALL X'S, TYPE 0\n";
print "(ZERO) AND TO START OVER IN THE MIDDLE OF A GAME, TYPE \n";
print "11 (ELEVEN).\n\n";

sub initialize
{
    my @a;
    print "1 2 3 4 5 6 7 8 9 10\n";
    print "X X X X X X X X X X\n";
    for my $i (0 .. 10) { $a[$i] = "X"; } # make sure [0] has a value just in case
    return @a;
}

while (1)
{
    my $Q = rand(1);
    my $C = 0;
    print "HERE IS THE STARTING LINE OF X'S.\n\n";
    my @A = initialize();

    while (1)
    {
        my $M = 0;
        my $N;
        while (1)
        {
            print "\nINPUT THE NUMBER: ";
            chomp($N = <>);
            if ($N != int($N) || $N < 0 ||  $N > 11)
            {
                print "ILLEGAL ENTRY--TRY AGAIN.\n";
                next;
            }
            last;
        }
        if ($N == 11) # start a new game
        {
            print "\n\n";
            last;
        }
        if ($N == 0) # reset line
        {
            @A = initialize();
            next;
        }

        if ($M != $N)
        {
            $M = $N;
            $A[$N] = ($A[$N] eq "O") ? "X" : "O";
            while ($M == $N)
            {
                my $R = tan($Q + $N / $Q - $N) - sin($Q / $N) + 336 * sin(8 * $N);
                $N = $R - int($R);
                $N = int(10 * $N);
                if ($A[$N] eq "O")
                {
                    $A[$N] = "X";
                    next;
                }
                $A[$N] = "O";
                last; # GOTO 610

                $A[$N] = "X";
            }
        }
        else
        {
            if ($A[$N] ne "O") { $A[$N] = "O"; }
            while ($M == $N)
            {
                my $R = .592 * (1 / tan($Q / $N + $Q)) / sin( $N * 2 + $Q) - cos($N);
                $N = $R - int($R);
                $N = int(10 * $N);
                if ($A[$N] eq "O")
                {
                    $A[$N] = "X";
                    next;
                }
                $A[$N] = "O";
                last;
            }
        }

        print "1 2 3 4 5 6 7 8 9 10\n";
        for my $i (1 .. 10) { print "$A[$i] "; }
        print "\n";
        $C++;
        my $i;
        for ($i=1 ; $i <= 10 ; $i++) {
            last if ($A[$i] ne "O");
        }
        if ($i == 11)
        {
            if ($C <= 12) { print "VERY GOOD.  YOU GUESSED IT IN ONLY $C GUESSES.\n"; }
            else          { print "TRY HARDER NEXT TIME.  IT TOOK YOU $C GUESSES.\n"; }
            last;
        }
    }
    print "DO YOU WANT TO TRY ANOTHER PUZZLE (Y/N): ";
    $_ = <>;
    print "\n";
    last if (m/^n/i);
}
